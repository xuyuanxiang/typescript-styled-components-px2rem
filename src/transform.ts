import * as ts from 'typescript';
import { replace } from './replace';
import configuration from './configuration';
import { createToken, isArrowFunction, isPropertyAccessExpression } from 'typescript';
import createPx2rem from './px2rem';

let _px2rem: ts.Identifier | undefined;
let _used: boolean = false;

function isStyledTagged(node: ts.TaggedTemplateExpression): boolean {
  const tag = node.tag;
  if (ts.isIdentifier(tag)) {
    return isStyled(tag);
  } else if (ts.isCallExpression(tag)) {
    if (ts.isIdentifier(tag.expression)) {
      return isStyled(tag.expression);
    } else if (ts.isPropertyAccessExpression(tag.expression)) {
      return isStyledMember(tag.expression);
    }
  } else if (ts.isPropertyAccessExpression(tag)) {
    return isStyledMember(tag);
  }
  return false;
}

function isStyledMember(node: ts.PropertyAccessExpression): boolean {
  if (ts.isIdentifier(node.expression)) {
    return isStyled(node.expression);
  } else if (ts.isPropertyAccessExpression(node.expression)) {
    return isStyledMember(node.expression);
  }
  return false;
}

function isStyled(id: ts.Identifier): boolean {
  const tags = configuration.config.tags;
  return Array.isArray(tags) && tags.includes(id.text);
}

function createIfDifference<T extends ts.Node>(text: string, create: (replaced: string) => T, origin: T): T {
  const replaced = replace(text);
  if (replaced !== text) {
    const newNode = create(replaced);
    newNode.parent = origin.parent;
    return newNode;
  }
  return origin;
}

function isPureExpression(
  expression: ts.Node,
): expression is
  | ts.LiteralExpression
  | ts.Identifier
  | ts.PropertyAccessExpression
  | ts.ElementAccessExpression
  | ts.CallExpression
  | ts.BinaryExpression {
  return (
    ts.isIdentifier(expression) ||
    ts.isLiteralExpression(expression) ||
    ts.isPropertyAccessExpression(expression) ||
    ts.isElementAccessExpression(expression) ||
    ts.isCallExpression(expression) ||
    ts.isBinaryExpression(expression)
  );
}

function createCallPx2rem(px2rem: ts.Identifier, ...expression: ts.Expression[]): ts.CallExpression {
  _used = true;
  return ts.createCall(px2rem, undefined, expression);
}

function createTemplateSpanExpressionVisitor(context: ts.TransformationContext, px2rem: ts.Identifier): ts.Visitor {
  const visitor: ts.Visitor = node => {
    if (isArrowFunction(node)) {
      if (ts.isBlock(node.body)) {
        return ts.updateArrowFunction(
          node,
          node.modifiers,
          node.typeParameters,
          node.parameters,
          node.type,
          node.equalsGreaterThanToken,
          ts.createCall(px2rem, undefined, [
            ts.createArrowFunction(
              undefined,
              undefined,
              [],
              undefined,
              createToken(ts.SyntaxKind.EqualsGreaterThanToken),
              node.body,
            ),
          ]),
        );
      } else {
        return ts.updateArrowFunction(
          node,
          node.modifiers,
          node.typeParameters,
          node.parameters,
          node.type,
          node.equalsGreaterThanToken,
          createCallPx2rem(px2rem, node.body),
        );
      }
    } else if (ts.isConditionalExpression(node)) {
      if (isPureExpression(node.whenTrue) && isPureExpression(node.whenFalse)) {
        return ts.updateConditional(
          node,
          node.condition,
          node.questionToken,
          createCallPx2rem(px2rem, node.whenTrue),
          node.colonToken,
          createCallPx2rem(px2rem, node.whenFalse),
        );
      } else if (isPureExpression(node.whenTrue)) {
        node = ts.updateConditional(
          node,
          node.condition,
          node.questionToken,
          createCallPx2rem(px2rem, node.whenTrue),
          node.colonToken,
          node.whenFalse,
        );
      } else if (isPureExpression(node.whenFalse)) {
        node = ts.updateConditional(
          node,
          node.condition,
          node.questionToken,
          node.whenTrue,
          node.colonToken,
          createCallPx2rem(px2rem, node.whenFalse),
        );
      }
    } else if (ts.isFunctionExpression(node)) {
      return ts.createArrowFunction(
        undefined,
        undefined,
        [
          ts.createParameter(
            undefined,
            undefined,
            ts.createToken(ts.SyntaxKind.DotDotDotToken),
            'args',
            undefined,
            undefined,
          ),
        ],
        undefined,
        ts.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
        createCallPx2rem(px2rem, node, ts.createSpread(ts.createIdentifier('args'))),
      );
    }
    // else if (ts.isBinaryExpression(node)) {
    //   switch (node.operatorToken.kind) {
    //     case ts.SyntaxKind.AmpersandAmpersandToken:
    //       if (isPureExpression(node.right)) {
    //         return ts.updateBinary(node, node.left, createCallPx2rem(px2rem, node.right), node.operatorToken);
    //       }
    //       break;
    //     case ts.SyntaxKind.BarBarToken:
    //       if (isPureExpression(node.left) && isPureExpression(node.right)) {
    //         return ts.updateBinary(
    //           node,
    //           createCallPx2rem(px2rem, node.left),
    //           createCallPx2rem(px2rem, node.right),
    //           node.operatorToken,
    //         );
    //       } else if (isPureExpression(node.left)) {
    //         node = ts.updateBinary(node, createCallPx2rem(px2rem, node.left), node.right, node.operatorToken);
    //       } else if (isPureExpression(node.right)) {
    //         node = ts.updateBinary(node, node.left, createCallPx2rem(px2rem, node.right), node.operatorToken);
    //       }
    //       break;
    //     default:
    //       return createCallPx2rem(px2rem, node);
    //   }
    // }
    return ts.visitEachChild(node, visitor, context);
  };

  return visitor;
}

function createStyledVisitor(context: ts.TransformationContext): ts.Visitor {
  const visitor: ts.Visitor = node => {
    if (ts.isNoSubstitutionTemplateLiteral(node)) {
      return createIfDifference(node.text, replaced => ts.createNoSubstitutionTemplateLiteral(replaced), node);
    } else if (ts.isTemplateHead(node)) {
      return createIfDifference(node.text, replaced => ts.createTemplateHead(replaced), node);
    } else if (ts.isStringLiteral(node)) {
      return createIfDifference(node.text, replaced => ts.createStringLiteral(replaced), node);
    } else if (ts.isTemplateTail(node)) {
      return createIfDifference(node.text, replaced => ts.createTemplateTail(replaced), node);
    } else if (ts.isTemplateMiddle(node)) {
      return createIfDifference(node.text, replaced => ts.createTemplateMiddle(replaced), node);
    }

    if (configuration.config.transformRuntime && _px2rem) {
      if (ts.isTemplateSpan(node) && node.literal.text && node.literal.text.startsWith('px')) {
        const text = node.literal.text.replace(/^px/, '');
        if (isPureExpression(node.expression)) {
          node = ts.updateTemplateSpan(
            node,
            createCallPx2rem(_px2rem, node.expression),
            ts.isTemplateMiddle(node.literal) ? ts.createTemplateMiddle(text) : ts.createTemplateTail(text),
          );
        } else {
          node = ts.updateTemplateSpan(
            node,
            ts.visitNode(node.expression, createTemplateSpanExpressionVisitor(context, _px2rem)),
            ts.isTemplateMiddle(node.literal) ? ts.createTemplateMiddle(text) : ts.createTemplateTail(text),
          );
        }
      }
    }
    return ts.visitEachChild(node, visitor, context);
  };

  return visitor;
}

function isStyledFunction(call: ts.CallExpression): boolean {
  const expression = call.expression;
  if (isPropertyAccessExpression(expression)) {
    return isStyledMember(expression);
  } else if (ts.isCallExpression(expression)) {
    return isStyledFunction(expression);
  } else if (ts.isIdentifier(expression)) {
    return isStyled(expression);
  }
  return false;
}

export const transformerFactory: ts.TransformerFactory<ts.SourceFile> = context => {
  const styledVisitor = createStyledVisitor(context);
  const visitor: ts.Visitor = node => {
    if (ts.isTaggedTemplateExpression(node) && isStyledTagged(node)) {
      return ts.visitNode(node, styledVisitor);
    } else if (ts.isCallExpression(node) && isStyledFunction(node)) {
      return ts.visitNode(node, styledVisitor);
    }

    return ts.visitEachChild(node, visitor, context);
  };
  return node => {
    if (configuration.config.transformRuntime) {
      _px2rem = ts.createUniqueName('px2rem');
    } else {
      _px2rem = undefined;
    }
    _used = false;
    const sourceFile = ts.visitNode(node, visitor);
    if (_used && _px2rem) {
      return ts.updateSourceFileNode(sourceFile, [...sourceFile.statements, createPx2rem(_px2rem)]);
    }
    return sourceFile;
  };
};
