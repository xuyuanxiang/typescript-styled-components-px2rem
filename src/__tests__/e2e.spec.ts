import * as ts from 'typescript';
import { readFileSync } from 'fs';
import { join } from 'path';
import transform from '../';

describe('e2e', () => {
  it('should work', function() {
    const options = {
      rootValue: 100,
      unitPrecision: 5,
      minPixelValue: 0,
      multiplier: 1,
      tags: ['styled', 'css', 'createGlobalStyle', 'keyframes'],
      transformRuntime: false,
    };
    const transformed = ts.transpileModule(readFileSync(join(__dirname, 'case.txt'), 'utf8'), {
      fileName: 'case.txt',
      compilerOptions: {
        skipLibCheck: true,
        target: ts.ScriptTarget.ES2015,
        module: ts.ModuleKind.ES2015,
        strict: true,
        esModuleInterop: true,
      },
      transformers: { before: [transform(options)] },
    });
    expect(transformed.outputText).toMatchSnapshot();
  });

  it('should transform runtime', function() {
    const options = {
      rootValue: 100,
      unitPrecision: 5,
      minPixelValue: 0,
      multiplier: 1,
      tags: ['styled', 'css', 'createGlobalStyle', 'keyframes'],
      transformRuntime: true,
    };
    const transformed = ts.transpileModule(readFileSync(join(__dirname, 'case.txt'), 'utf8'), {
      fileName: 'case.txt',
      compilerOptions: {
        skipLibCheck: true,
        target: ts.ScriptTarget.ES2015,
        module: ts.ModuleKind.ES2015,
        strict: true,
        esModuleInterop: true,
      },
      transformers: { before: [transform(options)] },
    });
    expect(transformed.outputText).toMatchSnapshot();
  });

  it('should polyfill px2rem when really used', function() {
    const options = {
      rootValue: 100,
      unitPrecision: 5,
      minPixelValue: 0,
      multiplier: 1,
      tags: ['styled', 'css', 'createGlobalStyle', 'keyframes'],
      transformRuntime: true,
    };
    const transformed = ts.transpileModule('const foo = 1', {
      fileName: 'case.txt',
      compilerOptions: {
        skipLibCheck: true,
        target: ts.ScriptTarget.ES2015,
        module: ts.ModuleKind.ES2015,
        strict: true,
        esModuleInterop: true,
      },
      transformers: { before: [transform(options)] },
    });
    expect(transformed.outputText).toMatchSnapshot();
  });
});
