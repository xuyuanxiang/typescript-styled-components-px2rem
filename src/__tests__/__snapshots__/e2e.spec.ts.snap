// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e should polyfill px2rem when really used 1`] = `
"\\"use strict\\";
const foo = 1;
"
`;

exports[`e2e should transform runtime 1`] = `
"import styled, { css, createGlobalStyle, keyframes } from 'styled-components';
const Animation = keyframes \`
  from {
    transform: translateX(1rem);
  }

  to {
    transform: translateX(-1rem);
  }
\`;
export const FunctionExpression = styled.button \`
  width: \${(...args) => px2rem_1(function (props) {
    return props.width;
}, ...args)}; /* Block Body */
  \${props => (props.disabled ? \\"height: 4rem\\" : \\"height: 2rem\\")};
\`;
const height = '44';
export const ArrowFunction = styled.input.attrs(props => ({
    type: 'password',
    size: props.size || \\"0.16rem\\",
    width: props.width || 100,
})) \`
  color: palevioletred;
  font-size: 0.14rem;
  border: 0.01rem solid palevioletred;
  border-radius: 0.08rem;
  width: \${props => px2rem_1(props.width)}; /* PropertyAccess Body */
  height: \${() => px2rem_1(height)}; /* Identifier Body */
  line-height: \${() => px2rem_1('44')}; /* StringLiteral Body */
  margin: \${() => px2rem_1(32)}; /* NumericLiteral Body */
  padding: \${props => props.size};
\`;
export const ArrowFunctionWithBlockBody = styled.button \`
  width: \${props => px2rem_1(() => {
    if (props.width) {
        return props.width;
    }
    else {
        return 0;
    }
})}; /* Block Body */
  \${props => (props.disabled ? \\"height: 4rem\\" : \\"height: 2rem\\")};
\`;
export const ArrowFunctionWithBinaryBody = styled.button \`
  \${props => props.disabled && \`
    width: 2rem;
    font-size: 0.14rem;
  \`};
  height: \${props => px2rem_1(!props.disabled && props.height)}; /* Binary Body */
\`;
export const ArrowFunctionWithConditionalBody = styled.button \`
  height: \${props => px2rem_1((props.height ? height : 100))}; /* Conditional Body */
\`;
const fontSize = 18;
function getHeight() {
    const height = 100;
    return height / 2;
}
const mixins = css \`
  padding: 0 0.16rem;
  margin: 0.16rem 0.32rem 0.16rem 0.32rem;
\`;
export const GlobalStyle = createGlobalStyle \`
  html body {
    \${mixins};
    font-size: \${px2rem_1(fontSize)}; /* Identifier */
    width: 10.24rem;
    height: \${px2rem_1(getHeight())}; /* CallExpression */
  }
\`;
export const StyledButton = styled.button \`
  width: 1.2rem;
  height: 0.32rem;
  font-size: 0.14rem;
\`;
export const ExtendStyledButton = styled(StyledButton)(({ padding }) => \`
  padding-left: 0.15rem;
  \${padding
    ? \`
      padding-left: 0.25rem;
  \` : ''}
\`);
export const PropertyAccessExpression = styled.button(props => \`
  display: inline;
  width: \${px2rem_1(props.width)};
  height: \${props.height};
  font-size: 0.16rem;
\`);
export const ThemeConsumer = styled.div \`
  font-size: \${props => px2rem_1(props.theme.fontSize)};
  color: \${props => props.theme.color};
\`;
export const ConditionalExpression = function ({ fontSize }) {
    const StyledButton = styled.button \`
    font-size: \${typeof fontSize === 'number' ? px2rem_1(fontSize) : props => px2rem_1(props === null || props === void 0 ? void 0 : props.theme.fontSize)};
  \`;
    return />;;
};
export const ConditionalExpressionWhenTrue = function ({ fontSize }) {
    const StyledButton = styled.button \`
    font-size: \${typeof fontSize !== 'number' ? props => px2rem_1(props === null || props === void 0 ? void 0 : props.theme.fontSize) : px2rem_1(fontSize)};
  \`;
    return />;;
};
export const ConditionalExpressionWhenFalse = function ({ fontSize }) {
    const StyledButton = styled.button \`
    font-size: \${typeof fontSize === 'number' ? px2rem_1(fontSize) : px2rem_1(16)};
  \`;
    return />;;
};
const condition = false;
function calc() {
    return 20;
}
export const BinaryExpression = styled.button \`
  \${condition || \`
    width: 2rem;
  \`};
  height: \${px2rem_1(condition || 100)};
  padding: \${px2rem_1(40 + 50)} 0.18rem \${px2rem_1('4')} 0.12rem;
  line-height: \${px2rem_1(calc() - 2)};
\`;
export const ConditaionalExpression = styled.div(props => \`
  flex: 0 0 30%;
  padding-left: 0.15rem;
  \${props.icon
    ? \`
     background-position: 0.09rem center;
     background-repeat: no-repeat;
     background-size: 0.09rem 0.15rem;
     padding-left: 0.25rem;
  \`
    : ''}
\`);
export const Foo = (function () { }) \`
foo
\`;
const Bar = [].foo \`
 bar
\`;
const ArrayExpressionCallee = [styled](props => \`
  color: palevioletred;
  font-size: 14px;
  border: 1px solid palevioletred;
  border-radius: 8px;
  width: \${props.width}px;
\`);
function px2rem_1(input, ...args) { if (typeof input === \\"function\\")
    return px2rem_1(input(...args), ...args); var value = parseFloat(input); var pixels = Number.isNaN(value) ? 0 : value; if (Math.abs(pixels) < 0)
    return \`\${pixels}px\`; var multiplier = Math.pow(10, 5 + 1); var wholeNumber = Math.floor((pixels * 1) / 100 * multiplier); return \`\${(Math.round(wholeNumber / 10) * 10) / multiplier}rem\`; }
"
`;

exports[`e2e should work 1`] = `
"import styled, { css, createGlobalStyle, keyframes } from 'styled-components';
const Animation = keyframes \`
  from {
    transform: translateX(1rem);
  }

  to {
    transform: translateX(-1rem);
  }
\`;
export const FunctionExpression = styled.button \`
  width: \${function (props) {
    return props.width;
}}px; /* Block Body */
  \${props => (props.disabled ? \\"height: 4rem\\" : \\"height: 2rem\\")};
\`;
const height = '44';
export const ArrowFunction = styled.input.attrs(props => ({
    type: 'password',
    size: props.size || \\"0.16rem\\",
    width: props.width || 100,
})) \`
  color: palevioletred;
  font-size: 0.14rem;
  border: 0.01rem solid palevioletred;
  border-radius: 0.08rem;
  width: \${props => props.width}px; /* PropertyAccess Body */
  height: \${() => height}px; /* Identifier Body */
  line-height: \${() => '44'}px; /* StringLiteral Body */
  margin: \${() => 32}px; /* NumericLiteral Body */
  padding: \${props => props.size};
\`;
export const ArrowFunctionWithBlockBody = styled.button \`
  width: \${props => {
    if (props.width) {
        return props.width;
    }
    else {
        return 0;
    }
}}px; /* Block Body */
  \${props => (props.disabled ? \\"height: 4rem\\" : \\"height: 2rem\\")};
\`;
export const ArrowFunctionWithBinaryBody = styled.button \`
  \${props => props.disabled && \`
    width: 2rem;
    font-size: 0.14rem;
  \`};
  height: \${props => !props.disabled && props.height}px; /* Binary Body */
\`;
export const ArrowFunctionWithConditionalBody = styled.button \`
  height: \${props => (props.height ? height : 100)}px; /* Conditional Body */
\`;
const fontSize = 18;
function getHeight() {
    const height = 100;
    return height / 2;
}
const mixins = css \`
  padding: 0 0.16rem;
  margin: 0.16rem 0.32rem 0.16rem 0.32rem;
\`;
export const GlobalStyle = createGlobalStyle \`
  html body {
    \${mixins};
    font-size: \${fontSize}px; /* Identifier */
    width: 10.24rem;
    height: \${getHeight()}px; /* CallExpression */
  }
\`;
export const StyledButton = styled.button \`
  width: 1.2rem;
  height: 0.32rem;
  font-size: 0.14rem;
\`;
export const ExtendStyledButton = styled(StyledButton)(({ padding }) => \`
  padding-left: 0.15rem;
  \${padding
    ? \`
      padding-left: 0.25rem;
  \` : ''}
\`);
export const PropertyAccessExpression = styled.button(props => \`
  display: inline;
  width: \${props.width}px;
  height: \${props.height};
  font-size: 0.16rem;
\`);
export const ThemeConsumer = styled.div \`
  font-size: \${props => props.theme.fontSize}px;
  color: \${props => props.theme.color};
\`;
export const ConditionalExpression = function ({ fontSize }) {
    const StyledButton = styled.button \`
    font-size: \${typeof fontSize === 'number' ? fontSize : props => props === null || props === void 0 ? void 0 : props.theme.fontSize}px;
  \`;
    return />;;
};
export const ConditionalExpressionWhenTrue = function ({ fontSize }) {
    const StyledButton = styled.button \`
    font-size: \${typeof fontSize !== 'number' ? props => props === null || props === void 0 ? void 0 : props.theme.fontSize : fontSize}px;
  \`;
    return />;;
};
export const ConditionalExpressionWhenFalse = function ({ fontSize }) {
    const StyledButton = styled.button \`
    font-size: \${typeof fontSize === 'number' ? fontSize : 16}px;
  \`;
    return />;;
};
const condition = false;
function calc() {
    return 20;
}
export const BinaryExpression = styled.button \`
  \${condition || \`
    width: 2rem;
  \`};
  height: \${condition || 100}px;
  padding: \${40 + 50}px 0.18rem \${'4'}px 0.12rem;
  line-height: \${calc() - 2}px;
\`;
export const ConditaionalExpression = styled.div(props => \`
  flex: 0 0 30%;
  padding-left: 0.15rem;
  \${props.icon
    ? \`
     background-position: 0.09rem center;
     background-repeat: no-repeat;
     background-size: 0.09rem 0.15rem;
     padding-left: 0.25rem;
  \`
    : ''}
\`);
export const Foo = (function () { }) \`
foo
\`;
const Bar = [].foo \`
 bar
\`;
const ArrayExpressionCallee = [styled](props => \`
  color: palevioletred;
  font-size: 14px;
  border: 1px solid palevioletred;
  border-radius: 8px;
  width: \${props.width}px;
\`);
"
`;
